vsintegration/src/FSharp.Editor/CodeLens/LineLens.fs

```diff
@@ -0,0 +1,697 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace rec Microsoft.VisualStudio.FSharp.Editor
+
+
+open System
+open System.Windows.Controls
+open System.Windows.Media
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Text.Editor
+open Microsoft.VisualStudio.Text.Formatting
+open System.ComponentModel.Composition
+open Microsoft.VisualStudio.Utilities
+open Microsoft.CodeAnalysis
+open System.Threading
+open Microsoft.FSharp.Compiler.SourceCodeServices
+open Microsoft.VisualStudio.Shell
+open Microsoft.VisualStudio
+open Microsoft.VisualStudio.LanguageServices
+open System.Windows
+open System.Collections.Generic
+open Microsoft.FSharp.Compiler.Range
+open Microsoft.FSharp.Compiler
+open Microsoft.FSharp.Compiler.Ast
+open Microsoft.CodeAnalysis.Editor.Shared.Extensions
+open Microsoft.CodeAnalysis.Editor.Shared.Utilities
+open Microsoft.CodeAnalysis.Classification
+open Internal.Utilities.StructuredFormat
+open Microsoft.VisualStudio.Text.Tagging
+open System.Collections.Concurrent
+open System.Collections
+open System.Windows.Media.Animation
+open System.Globalization
+
+open Microsoft.VisualStudio.FSharp.Editor.Logging
+open Microsoft.CodeAnalysis.Text
+open System
+
+type internal LineLensDisplayService (view, buffer) as self =
+    // Custom simple tagger logic
+    inherit CodeLensDisplayService(view, buffer)
+    /// Saves the ui context to switch context for ui related work.
+    let uiContext = SynchronizationContext.Current
+
+    /// The tags changed event to notify if the data for the tags has changed.
+    let bufferChangedEvent = new Event<EventHandler<TextContentChangedEventArgs>,TextContentChangedEventArgs>()
+
+    // Tracks the created ui elements per TrackingSpan
+    let uiElements = Dictionary<_,_>()
+    /// Caches the current used trackingSpans per line. One line can contain multiple trackingSpans
+    let mutable trackingSpans = Dictionary<_, Generic.List<_>>()
+    /// Text view for accessing the adornment layer.
+    let mutable view: IWpfTextView = view
+    /// The line lens layer for adding and removing adornments.
+    let mutable lineLensLayer: IAdornmentLayer = view.GetAdornmentLayer "LineLens"
+    /// Tracks the recent first + last visible line numbers for adornment layout logic.
+    let mutable recentFirstVsblLineNmbr, recentLastVsblLineNmbr = 0, 0
+    /// Tracks the adornments on the layer.
+    let mutable addedAdornments = HashSet()
```

Comment:
Why mutable? Is never reset

Reply:
Only line 57 needs to be mutable, you're right. Thanks again.